base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,811 @@
+        GetRegString(key, WinVal, Output) && !Output.IsEmpty())
+    {
+        return TRUE;
+    }
+    WCHAR windir[MAX_PATH];
+    GetWindowsDirectoryW(windir, _countof(windir));
+    CStringW path = BuildPath(BuildPath(windir, L"inf"), RosInf), section;
+    DWORD lang = 0, lctype = LOCALE_ILANGUAGE | LOCALE_RETURN_NUMBER;
+    if (GetLocaleInfoW(GetUserDefaultLCID(), lctype, (LPWSTR) &lang, sizeof(lang) / sizeof(WCHAR)))
+    {
+        section.Format(L"Strings.%.4x", lang);
+        if (ReadIniValue(path, section, RosVal, Output) > 0)
+            return TRUE;
+        section.Format(L"Strings.%.2x", PRIMARYLANGID(lang));
+        if (ReadIniValue(path, section, RosVal, Output) > 0)
+            return TRUE;
+    }
+    return ReadIniValue(path, L"Strings", RosVal, Output) > 0;
+}
+
+static BOOL
+CreateShortcut(const CStringW &Target)
+{
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    UINT csidl = Info.PerUser ? CSIDL_PROGRAMS : CSIDL_COMMON_PROGRAMS;
+    CStringW rel = Info.ShortcutFile, path, dir, tmp;
+
+    if (FAILED(GetSpecialPath(csidl, path, Info.GetGuiOwner())))
+        return TRUE; // Pretend everything is OK
+
+    int cat;
+    if (Info.Parser.GetInt(DB_CATEGORY, cat) && cat == ENUM_CAT_GAMES)
+    {
+        if (GetLocalizedSMFolderName(L"SM_GamesName", L"shortcuts.inf", L"Games", tmp))
+        {
+            path = BuildPath(path, tmp);
+        }
+    }
+
+    rel.Replace('/', '\\');
+    path = BuildPath(path, rel.GetString());
+    UINT SHPPFW = SHPPFW_DIRCREATE | SHPPFW_IGNOREFILENAME;
+    HRESULT hr = SHPathPrepareForWriteW(Info.GetGuiOwner(), NULL, path, SHPPFW);
+    if ((Info.Error = ErrorFromHResult(hr)) != 0)
+    {
+        ErrorBox(Info.Error);
+        return FALSE;
+    }
+
+    CComPtr<IShellLinkW> link;
+    hr = link.CoCreateInstance(CLSID_ShellLink, IID_IShellLinkW);
+    if (SUCCEEDED(hr))
+    {
+        if (SUCCEEDED(hr = link->SetPath(Target)))
+        {
+            if (SUCCEEDED(GetCustomIconPath(Info, tmp)))
+            {
+                LPWSTR p = tmp.GetBuffer();
+                int idx = PathParseIconLocation(p);
+                link->SetIconLocation(p, idx);
+            }
+            CComPtr<IPersistFile> persist;
+            if (SUCCEEDED(hr = link->QueryInterface(IID_IPersistFile, (void**)&persist)))
+            {
+                hr = persist->Save(path, FALSE);
+            }
+        }
+    }
+    if (SUCCEEDED(hr))
+    {
+        if (AddEntry(UNOP_FILE, path))
+        {
+            SplitFileAndDirectory(path, &dir);
+            AddEntry(UNOP_EMPTYDIR, dir);
+        }
+    }
+    else
+    {
+        ErrorBox(ErrorFromHResult(hr));
+    }
+    return !Info.Error;
+}
+
+static BOOL
+InstallFiles(const CStringW &SourceDirBase, const CStringW &Spec,
+             const CStringW &DestinationDir)
+{
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    CStringW sourcedir, filespec;
```

Comment:
```suggestion
    BOOL failed = FALSE;

    CStringW sourcedir, filespec;
```

Reply:
Also, what about using instead a `BOOL Success` so that you don't have to always negate these things? `return !failed;` --> `return Success;`, and `failed = !Whatever();` --> `Success = Whatever();` ?